eventlog/service.go

```diff
@@ -119,6 +119,83 @@ func (s *Service) GetEvent(req *GetEventRequest) (*GetEventResponse, error) {
 	}, nil
 }
 
+var days = 24 * time.Hour
+var years = 365 * days
+
+// Search will search the event log for matching entries.
+func (s *Service) Search(req *SearchRequest) (*SearchResponse, error) {
+	if req.To == 0 {
+		req.To = time.Now().UnixNano()
+	}
+
+	v := s.omni.GetCollectionView(req.ID)
+
+	// Find the latest bucket
+	_, b, err := getLatestBucket(v)
+	if err == errIndexMissing {
+		// There are no events yet on this chain, so return no results.
+		return &SearchResponse{}, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	// bend is normally updated from the last bucket's start. For the latest
+	// bucket, bend is now.
+	bend := time.Now().UnixNano()
+
+	// Walk backwards until the current bucket's start is before our From
+	// or this is the first bucket. Remember the bucket id's that we'll be searching.
+
+	// Walk backwards in the bucket chain through 2 zones: first where the
+	// bucket covers time that is not in our search range, and then where the buckets
+	// do cover the search range.
+	var buckets []*bucket
+	for {
+		if req.From > bend {
+			// This bucket is before the search range, so we are done walking back the bucket chain.
+			break
+		}
+
+		if req.To < b.Start {
+			// This bucket is after the search range, so we do not add it to buckets, but
+			// we keep walking up the chain.
+		} else {
+			buckets = append(buckets, b)
+		}
+
+		if b.isFirst() {
+			break
+		}
+		bend = b.Start
+		b, err = getBucketByID(v, b.Prev)
+		if err != nil {
+			// We might have found some buckets to search, so exit the loop.
+			break
+		}
+	}
+
+	reply := &SearchResponse{
+		Truncated: false,
+	}
+	if len(buckets) == 0 {
+		return reply, nil
+	}
+
+	for _, b := range buckets {
+		for _, e := range b.EventRefs {
+			ev, err := getEventByID(v, e)
+			if err != nil {
+				return nil, err
```

Comment:
If we allow failures in `getBucketByID`, we should also allow failures in `getBucketByID`, i.e., we try to return as many events as we can and let the client know that some results might be missing via the `Truncated` flag. Alternatively, we immediately fail if either of these functions returns an error, because that would indicate the time buckets are not created correctly, which is a critical error to me. I'm leaning towards the latter.